Line/iOS/LINELoginDemo/LINELoginDemo/LineAuthManager.m

```diff
@@ -0,0 +1,230 @@
+    [self.topViewController dismissViewControllerAnimated:YES completion:nil];
+}
+
+- (void)lineAdapterAuthorizationDidChange:(NSNotification*)aNotification {
+    LineAdapter *_adapter = [aNotification object];
+    if ([_adapter isAuthorized]) {
+        if (![self.lineAdapter canAuthorizeUsingLineApp]) {
+            // Authenticated using Webview, so need to dismiss the webview controller
+            if (self.topViewController.presentedViewController) {
+                [self.topViewController dismissViewControllerAnimated:YES completion:nil];
+            }
+        }
+        
+        // Connection completed to LINE. Start step 2 validating LINE access token
+        NSString *lineAccessToken = self.lineAdapter.getLineApiClient.accessToken;
+//        NSLog(@"DEBUG: LINE Access Token = %@",lineAccessToken);
+        [self requestFirebaseAuthTokenWithLINEAccessToken:lineAccessToken];
+    } else {
+        // Return error to login callback
+        NSError *error = [[aNotification userInfo] objectForKey:@"error"];
+        if (error)
+        {
+            [self returnSignInResult:nil error:error];
+        }
+    }
+}
+
+#pragma mark - Step 2: Communication with own validation server
+- (void)requestFirebaseAuthTokenWithLINEAccessToken:(NSString *)lineAccessToken {
+    // Stop current Firebase token fetcher if there's any
+    if (self.firebaseTokenFetcher && self.firebaseTokenFetcher.isFetching) {
+        [self.firebaseTokenFetcher stopFetching];
+    }
+    
+    // STEP 2: Exchange LINE access token for Firebase Custom Auth token
+    NSString *urlString = [NSString stringWithFormat:@"%@/verifyToken", YOUR_VALIDATION_SERVER_DOMAIN];
+    NSURL *url = [NSURL URLWithString:urlString];
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
+    [request setHTTPMethod:@"POST"];
+    [request setValue:@"application/json" forHTTPHeaderField:@"content-type"];
+    
+    NSDictionary *token = @{@"token" : lineAccessToken};
+    NSError *error;
+    NSData *requestBody = [NSJSONSerialization dataWithJSONObject:token
+                                                          options:kNilOptions error:&error];
+    if (error) {
+        [self returnSignInResult:nil error:error];
+    }
+    [request setHTTPBody:requestBody];
+    
+    self.firebaseTokenFetcher = [GTMHTTPFetcher fetcherWithRequest:request];
+    self.firebaseTokenFetcher.allowedInsecureSchemes = @[@"http"]; // Allow fetching http so that you can run your server sample code locally without https
+    __weak typeof(self) wSelf = self;
+    
+    [self.firebaseTokenFetcher beginFetchWithCompletionHandler:^(NSData *data, NSError *error) {
+        if (error) {
+            [wSelf returnSignInResult:nil error:error];
+        } else {
+            NSError *jsonSerializationError;
+            NSDictionary *response = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonSerializationError];
+            
+            if (error) {
```

Comment:
Fixed in this [commit](https://github.com/firebase/custom-auth-samples/pull/1/commits/cfb5c9d622d615a16b4c27754be32543065571cc) :)